pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/AnnotationSuppressionUtil.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.java.rule;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import net.sourceforge.pmd.Rule;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.java.ast.ASTAnnotation;
+import net.sourceforge.pmd.lang.java.ast.ASTAnyTypeDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTFieldDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTFormalParameter;
+import net.sourceforge.pmd.lang.java.ast.ASTLiteral;
+import net.sourceforge.pmd.lang.java.ast.ASTLocalVariableDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTMethodOrConstructorDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTTypeDeclaration;
+import net.sourceforge.pmd.lang.java.ast.Annotatable;
+import net.sourceforge.pmd.lang.java.typeresolution.TypeHelper;
+
+/**
+ * Helper methods to suppress violations based on annotations.
+ *
+ * An annotation suppresses a rule if the annotation is a {@link SuppressWarnings},
+ * and if the set of suppressed warnings ({@link SuppressWarnings#value()})
+ * contains at least one of those:
+ * <ul>
+ * <li>"PMD" (suppresses all rules);
+ * <li>"PMD.rulename", where rulename is the name of the given rule;
+ * <li>"all" (conventional value to suppress all warnings).
+ * </ul>
+ *
+ * <p>Additionally, the following values suppress a specific set of rules:
+ * <ul>
+ * <li>{@code "unused"}: suppresses rules like UnusedLocalVariable or UnusedPrivateField;
+ * <li>{@code "serial"}: suppresses BeanMembersShouldSerialize and MissingSerialVersionUID;
+ * <li>TODO "fallthrough" #1899
+ * </ul>
+ */
+final class AnnotationSuppressionUtil {
```

Comment:
considering this is only used in `JavaRuleViolation`, do we really need it to be public API?

Reply:
It's not public API, it's package-private.

Reply:
I meant, moving it to an internal package to document it as something outside devs shouldn't use.

Reply:
Doesn't the type being package private do exactly that? IMHO it has the added benefit that it's impossible to use it outside of its intended scope, even in our own codebase. If a type doesn't need to be public then why would we artificially make it so?